pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go

```diff
@@ -86,13 +87,19 @@ func createKnativeService(componentMeta metav1.ObjectMeta,
 		annotations[autoscaling.ClassAnnotationKey] = autoscaling.KPA
 	}
 	lastRolledoutRevision := componentStatus.LatestRolledoutRevision
-	// This is to handle case when the latest ready revision is rolled out with 100% and then rolled back
-	// so here we need to get the revision that is previously rolled out with 100%
-	if componentStatus.LatestRolledoutRevision == componentStatus.LatestReadyRevision &&
-		componentStatus.LatestReadyRevision == componentStatus.LatestCreatedRevision &&
-		componentExtension.CanaryTrafficPercent != nil && *componentExtension.CanaryTrafficPercent < 100 {
+
+	// This is to handle case when the first revision rolled out.
+	if strings.HasSuffix(componentStatus.LatestRolledoutRevision, "-00001") {
```

Comment:
this might not be a reliable check as user can choose to override the revision name, I think it is fine to check if `LatestRolledoutRevision` is not empty.

also why do we still need `lastRolledoutRevision = componentStatus.PreviousRolledoutRevision` ?

Reply:
`lastRolledoutRevision = componentStatus.PreviousRolledoutRevision` is actually for set `lastRolledoutRevision` to empty. When I roll out the first revision with 1 traffic percent, The revision status change log is as follows:

```
{"level":"info","ts":1647955245.7690232,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"","LatestCreatedRevision":"","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955245.9323924,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"","LatestCreatedRevision":"","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955246.1248095,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"","LatestCreatedRevision":"","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955246.2239192,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"","LatestCreatedRevision":"","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955246.240706,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
```

```
{"level":"info","ts":1647955295.2115061,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955295.2362065,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955295.3836544,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955295.4536629,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955295.5241714,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955295.5614073,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955295.8245306,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955295.924045,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955295.9450102,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955295.9623299,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.0252357,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.1252217,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.1559985,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.2245576,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.2484438,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.3239365,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.3661742,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.7657032,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.8372052,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.9283204,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955296.9545665,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955297.1925607,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955297.2462313,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955297.4247804,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955297.5241365,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955297.7251525,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955297.749457,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955297.933567,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955297.964174,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955298.2083428,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955298.2352567,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955298.2509632,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955298.3649652,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955298.3926592,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955298.5239766,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955298.63417,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955298.7239954,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955298.7527847,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.1775756,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.252463,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.3240964,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.3448806,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.372835,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.4322815,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.4523559,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.6674411,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.7398708,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.823951,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.8468132,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955299.8683274,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.0341926,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.0640078,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.139318,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.185193,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.23705,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.4156194,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.448103,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.5240514,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.624298,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.655463,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.8565493,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.8808355,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.8999777,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.9406445,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955300.9859104,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955301.0343618,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
{"level":"info","ts":1647955301.0574412,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":0}
```
When isvc status is ready, the status starts to change constantly
```
NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION               LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.amh-group.com   False   100    0        test-pmml-predictor-default-00001   test-pmml-predictor-default-00001   70s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.amh-group.com   False          100                              test-pmml-predictor-default-00001   71s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION               LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.amh-group.com   False   100    0        test-pmml-predictor-default-00001   test-pmml-predictor-default-00001   73s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION               LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.amh-group.com   True    100    0        test-pmml-predictor-default-00001   test-pmml-predictor-default-00001   75s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.amh-group.com   True 
``` 
If I remove the if condition and not set `lastRolledoutRevision` to empty, how can I solve this problem?

Reply:
> this might not be a reliable check as user can choose to override the revision name, I think it is fine to check if `LatestRolledoutRevision` is not empty.
> 
> also why do we still need `lastRolledoutRevision = componentStatus.PreviousRolledoutRevision` ?

Can we overwrite the revision name by isvc? If not I think we do not have to consider this condition,  at least not particularly necessary at the moment.

Reply:
I thought we talked about that the canaryTrafficPercent 1 does not really make sense for the first revision rollout, we can override that to 100 and only allowing it for the second revision rollout. Any problem you see from the following logic?

```golang
if componentStatus.LatestRolledoutRevision == "" && canaryTrafficPercent != nil {
    canaryTrafficPercent = 100
}
```

also I would not rely on the revision name to check the first revision, knative can change that and in fact in older release it is a generated name also knative allows overriding that and we just did not expose it in kserve.

Reply:
```
if componentStatus.LatestRolledoutRevision == "" && canaryTrafficPercent != nil {
    canaryTrafficPercent = 100
}
```
With the logic, isvc changes as follows:
```
NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION               LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.wenyangchou.com   True    99     1        test-pmml-predictor-default-00001   test-pmml-predictor-default-00001   62s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION               LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.wenyangchou.com   True    99     1        test-pmml-predictor-default-00001   test-pmml-predictor-default-00001   63s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.wenyangchou.com   True           100                              test-pmml-predictor-default-00001   64s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION               LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.wenyangchou.com   False   99     1        test-pmml-predictor-default-00001   test-pmml-predictor-default-00001   65s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION               LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.wenyangchou.com   True    99     1        test-pmml-predictor-default-00001   test-pmml-predictor-default-00001   67s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.wenyangchou.com   False          100                              test-pmml-predictor-default-00001   68s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION               LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.wenyangchou.com   True    99     1        test-pmml-predictor-default-00001   test-pmml-predictor-default-00001   70s

NAME        URL                                                          READY   PREV   LATEST   PREVROLLEDOUTREVISION               LATESTREADYREVISION                 AGE
test-pmml   http://test-pmml.phantom.dev-predict-phantom.wenyangchou.com   True    99     1        test-pmml-predictor-default-00001   test-pmml-predictor-default-00001   72s
```
revision status changes :
```
{"level":"info","ts":1647998486.6303008,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":1}
{"level":"info","ts":1647998486.7241223,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":1}
{"level":"info","ts":1647998486.9247065,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":1}
{"level":"info","ts":1647998486.9527392,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":1}
{"level":"info","ts":1647998487.024337,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":100}
{"level":"info","ts":1647998487.0524247,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":100}
{"level":"info","ts":1647998487.078394,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":100}
{"level":"info","ts":1647998487.2460775,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":100}
{"level":"info","ts":1647998487.2860572,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":100}
{"level":"info","ts":1647998487.3398628,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":1}
{"level":"info","ts":1647998487.5762186,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":1}
{"level":"info","ts":1647998487.638925,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":1}
{"level":"info","ts":1647998487.8716,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":1}
{"level":"info","ts":1647998487.9417357,"logger":"KsvcReconciler","msg":"revision status:","LatestRolledoutRevision":"test-pmml-predictor-default-00001","LatestReadyRevision":"test-pmml-predictor-default-00001","LatestCreatedRevision":"test-pmml-predictor-default-00001","PreviousRolledoutRevision":"","CanaryTrafficPercent":1}
```
If I do not set the previous revision, the latest revision is the same as the previous revision and compete with each other because the isvc will trigger asynchronously.

Reply:
I can see the issue now, initially [`LatestRolledoutRevision` was empty](https://github.com/kserve/kserve/blob/master/pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go#L118) so in the reconcile step it updates traffic split to 100 for the revision 1, then `LatestRolledoutRevision` was updated to revision 1 in status propagation, the next reconcile step updates traffic split to 1/99 for revision 1 because [`LatestRolledoutRevision` is not empty and `canaryTrafficPercent` is set](https://github.com/kserve/kserve/blob/master/pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go#L98), finally in the status propagation it [rolls back `LatestRolledoutRevision` to `PreviousRolledoutRevision`](https://github.com/kserve/kserve/blob/master/pkg/apis/serving/v1beta1/inference_service_status.go#L215) as the latest revision split is < 100 which results in infinite loop.

Reply:
the problem with your check is that it prevents doing canary rollout for the second revision, so let's think what would be the proper fix.

Reply:
I think really solve this problem, we need to add a status to mark whether the current revision is before or after `updating knative`. The revision can only be used for judgment after `updating knative`.

Reply:
We could store the `generation` field in the isvc status,  every time isvc status propagation is invoked from ksvc status update the controller updates the isvc status `generation` field.

Reply:
@wenyangchou The fix should work for all the cases without spurious revision spawn up, let's get in this first and we can discuss with your idea if that can improve the current logic

Reply:
There are still some preblems. When I roll out the first revision with 100 traffic percent, and then roll out the first revision with 0 traffic percent, the isvc status would always be false.

Reply:
@wenyangchou It should be fixed for your case now with https://github.com/kserve/kserve/pull/2097/commits/8eb4ccd85dd364df2753933420d07975792ad0f9

Reply:
approve

Reply:
approve